modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/TxFinishReplicaRequestHandler.java

```diff
@@ -0,0 +1,337 @@
+ */
+public class TxFinishReplicaRequestHandler {
+    private static final IgniteLogger LOG = Loggers.forClass(TxFinishReplicaRequestHandler.class);
+
+    /** Factory to create RAFT command messages. */
+    private static final PartitionReplicationMessagesFactory PARTITION_REPLICATION_MESSAGES_FACTORY =
+            new PartitionReplicationMessagesFactory();
+
+    /** Factory for creating replica command messages. */
+    private static final ReplicaMessagesFactory REPLICA_MESSAGES_FACTORY = new ReplicaMessagesFactory();
+
+    private final TxStatePartitionStorage txStatePartitionStorage;
+    private final ClockService clockService;
+    private final TxManager txManager;
+    private final ReplicationGroupId replicationGroupId;
+
+    private final SchemaCompatibilityValidator schemaCompatValidator;
+    private final ReliableCatalogVersions reliableCatalogVersions;
+    private final ReplicationRaftCommandApplicator raftCommandApplicator;
+    private final ReplicaTxFinisher replicaTxFinisher;
+
+
+    /** Constructor. */
+    public TxFinishReplicaRequestHandler(
+            TxStatePartitionStorage txStatePartitionStorage,
+            ClockService clockService,
+            TxManager txManager,
+            ValidationSchemasSource validationSchemasSource,
+            SchemaSyncService schemaSyncService,
+            CatalogService catalogService,
+            RaftCommandRunner raftCommandRunner,
+            ReplicationGroupId replicationGroupId
+    ) {
+        this.txStatePartitionStorage = txStatePartitionStorage;
+        this.clockService = clockService;
+        this.txManager = txManager;
+        this.replicationGroupId = replicationGroupId;
+
+        schemaCompatValidator = new SchemaCompatibilityValidator(validationSchemasSource, catalogService, schemaSyncService);
+        reliableCatalogVersions = new ReliableCatalogVersions(schemaSyncService, catalogService);
+        raftCommandApplicator = new ReplicationRaftCommandApplicator(raftCommandRunner, replicationGroupId);
+        replicaTxFinisher = new ReplicaTxFinisher(txManager);
+    }
+
+    /**
+     * Processes transaction finish request.
+     * <ol>
+     *     <li>Get commit timestamp from finish replica request.</li>
+     *     <li>If attempting a commit, validate commit (and, if not valid, switch to abort)</li>
+     *     <li>Run specific raft {@code FinishTxCommand} command, that will apply txn state to corresponding txStateStorage.</li>
+     *     <li>Send cleanup requests to all enlisted primary replicas.</li>
+     * </ol>
+     *
+     * @param request Transaction finish request.
+     * @return future result of the operation.
+     */
+    public CompletableFuture<TransactionResult> processTxFinishAction(TxFinishReplicaRequest request) {
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-19170 Use ZonePartitionIdMessage and remove cast
+        Map<TablePartitionId, String> enlistedGroups = asTablePartitionIdStringMap(request.groups());
```

Comment:
That doesn't look correct to me. Within colocation we need both tableIds and zonesIds. Latter in order to check whether primaries were not changed - validateCommit and to group TxCleanupMessage on per-node basis. How it's implemented?

Reply:
As I can see in the code, `enlistedGroups` are currently used for:

1. Table schema change validation (this will continue to be based on table IDs)
2. These groups are stored in `TxMeta` where they later get used for cleanup (not implemented in this PR, except for unlocking where these groups are not used) and WI resolution (not implemented here as well).
3. For cleanup (only unlocking is currently implemented)

WI resolution is still using the old mechanics, and I believe we don't have any tests for zone-based WI switch yet.

Reply:
>Latter in order to check whether primaries were not changed

Where do we do it?

Reply:
> For cleanup (only unlocking is currently implemented)

Yep, and in order to release locks it's required to send corresponding messages to zone primaries in case of collocation track.

Long story short, in order to validate schema and in order to perform WI conversion tableIds are required. For all other purposes I ZonePartitionId's should be used.

Reply:
Agreed to cover it in different ticket.